hail/src/main/scala/is/hail/expr/ir/Emit.scala

```diff
@@ -519,32 +559,25 @@ private class Emit(
                       .concat(s.pType.asInstanceOf[PString].loadString(coerce[Long](codeS.v)))))
               }
         }
-        val xma = mb.newLocal[Boolean]()
-        val xa = mb.newPLocal(pArray)
-        val xi = mb.newLocal[Int]
-        val len = mb.newLocal[Int]
-        val xmi = mb.newLocal[Boolean]()
-        val xmv = mb.newLocal[Boolean]()
-        val setup = Code(
-          codeA.setup,
-          xma := codeA.m,
-          codeI.setup,
-          xmi := codeI.m,
-          (xmi || xma).mux(
-            xmv := true,
-            Code(
-              xa := codeA.pv,
-              xi := coerce[Int](codeI.v),
-              len := xa.load().asIndexable.loadLength(),
-              (xi < len && xi >= 0).mux(
-                xmv := !xa.load().asIndexable.isElementDefined(xi),
-                Code._fatal[Unit](errorTransformer(
+        
+        EmitCode.fromI(mb) { cb =>
+          emit(a).toI(cb).flatMap(cb) { (ac) =>
+            emit(i).toI(cb).flatMap(cb) { (ic) =>
+              val av = ac.asIndexable.memoize(cb, "aref_a")
+              val iv = cb.memoize(ic.tcode[Int], "aref_i")
+
+              cb.ifx(iv < 0 || iv >= av.loadLength(), {
+                cb._fatal(errorTransformer(
                   const("array index out of bounds: index=")
-                    .concat(xi.load().toS)
+                    .concat(iv.toS)
                     .concat(", length=")
-                    .concat(len.load().toS)))))))
+                    .concat(av.loadLength().toS)))
+              })
+              av.loadElement(cb, iv)
```

Comment:
I ask because we memoize `iv` above, and then memoize it in `loadElement` as well.